client/src/telethon/_impl/mtsender/sender.py

```diff
@@ -192,62 +187,61 @@ async def connect(
         return cls(
             dc_id=dc_id,
             addr=addr,
-            lock=Lock(),
             _logger=base_logger.getChild("mtsender"),
+            _lock=Lock(),
             _reader=reader,
             _writer=writer,
             _transport=transport,
             _mtp=mtp,
-            _mtp_buffer=bytearray(),
-            _requests=[],
-            _request_event=Event(),
+            _requests=Queue(1),
             _next_ping=asyncio.get_running_loop().time() + PING_DELAY,
-            _read_buffer=bytearray(),
-            _write_drain_pending=False,
         )
 
     async def disconnect(self) -> None:
         self._writer.close()
         await self._writer.wait_closed()
 
-    def enqueue(self, request: RemoteCall[Return]) -> Future[bytes]:
-        rx = self._enqueue_body(bytes(request))
-        self._request_event.set()
+    async def enqueue(self, request: RemoteCall[Return]) -> Future[bytes]:
+        rx = await self._enqueue_body(bytes(request))
         return rx
 
     async def invoke(self, request: RemoteCall[Return]) -> bytes:
-        rx = self._enqueue_body(bytes(request))
+        rx = await self._enqueue_body(bytes(request))
         return await self._step_until_receive(rx)
 
     async def send(self, body: bytes) -> bytes:
-        rx = self._enqueue_body(body)
+        rx = await self._enqueue_body(body)
         return await self._step_until_receive(rx)
 
-    def _enqueue_body(self, body: bytes) -> Future[bytes]:
+    async def _enqueue_body(self, body: bytes) -> Future[bytes]:
         oneshot = asyncio.get_running_loop().create_future()
-        self._requests.append(Request(body=body, state=NotSerialized(), result=oneshot))
+
+        await self._requests.put(
+            Request(body=body, state=NotSerialized(), result=oneshot)
+        )
+
         return oneshot
 
     async def _step_until_receive(self, rx: Future[bytes]) -> bytes:
         while True:
+            # TODO: check step is running
             await self.step()
             if rx.done():
                 return rx.result()
 
     async def step(self) -> list[Updates]:
-        async with self.lock:
+        async with self._lock:
             return await self._step()
 
     async def _step(self) -> list[Updates]:
-        self._try_fill_write()
+        request = await self._requests.get()
+        await self._try_fill_write(request)
```

Comment:
Network should be able to run even if there are no requests to make. This will block until there is one which does not seem correct.